src/main/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentStepExecution.java

```diff
@@ -0,0 +1,183 @@
+import java.io.StringWriter;
+import java.util.*;
+
+public class SSHAgentStepExecution extends AbstractStepExecutionImpl {
+
+    /**
+     * Serial Version UID.
+     */
+    private static final long serialVersionUID = -1912198498615126332L;
+
+    @StepContextParameter
+    private transient TaskListener listener;
+
+    @StepContextParameter
+    private transient Run build;
+
+    @StepContextParameter
+    private transient Launcher launcher;
+
+    @Inject
+    private transient SSHAgentStep step;
+
+    /**
+     * The proxy for the real remote agent that is on the other side of the channel (as the agent needs to
+     * run on a remote machine)
+     */
+    private transient RemoteAgent agent = null;
+
+    @Override
+    public boolean start() throws Exception {
+        StepContext context = getContext();
+        String socket = initRemoteAgent();
+        context.newBodyInvoker().
+                withContext(EnvironmentExpander.merge(getContext().get(EnvironmentExpander.class), new ExpanderImpl(socket))).
+                withCallback(new Callback()).withDisplayName(null).start();
+        return false;
+    }
+
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        if (agent != null) {
+            agent.stop();
+            listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped());
+        }
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        try {
+            initRemoteAgent();
+        } catch (Exception e) {
+            listener.getLogger().println(Messages.SSHAgentBuildWrapper_CouldNotStartAgent());
+        }
+    }
+
+    private class Callback extends BodyExecutionCallback {
+
+        /**
+         * Serial Version UID.
+         */
+        private static final long serialVersionUID = 4118096102821683615L;
+
+        @Override
+        public void onSuccess(StepContext context, Object result) {
+            if (agent != null) {
+                agent.stop();
+                listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped());
+            }
+            getContext().onSuccess(result);
+        }
+
+        @Override
+        public void onFailure(StepContext context, Throwable t) {
+            if (agent != null) {
+                agent.stop();
+                listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped());
+            }
+            getContext().onFailure(t);
+        }
+
+    }
+
+    private static final class ExpanderImpl extends EnvironmentExpander {
+
+        /**
+         * Serial Version UID.
+         */
+        private static final long serialVersionUID = 1;
+
+        /**
+         * Set of environment variables to overrides (or add).
+         */
+        private final Map<String,String> overrides;
+
+        private ExpanderImpl(final String socket) {
+            this.overrides = new HashMap<String,String>();
+            overrides.put("SSH_AUTH_SOCK", socket);
```

Comment:
Silly to keep a `Map`-valued field here. Would suffice to keep the `String socket` as a field.